src/WebJobs.Script/OutOfProc/Rpc/JobHostLanguageWorkerChannelManager.cs

```diff
@@ -34,19 +41,48 @@ public void DisposeAndRemoveChannel(ILanguageWorkerChannel channel)
 
         public void DisposeAndRemoveChannels()
         {
+            Stopwatch timer = Stopwatch.StartNew();
             foreach (string channelId in _channels.Keys)
             {
                 if (_channels.TryRemove(channelId, out ILanguageWorkerChannel removedChannel))
                 {
-                    _logger.LogDebug("Disposing language worker channel with id:{workerId}", removedChannel.Id);
-                    (removedChannel as IDisposable)?.Dispose();
+                    if (removedChannel.FunctionInputBuffers.Values.Sum(val => val.Count) > 0)
+                    {
+                        using (ManualResetEventSlim mres1 = new ManualResetEventSlim(false))
+                        {
+                            IDisposable subscription = Observable
+                                                .Interval(TimeSpan.FromSeconds(1.0))
+                                                .Subscribe(x =>
+                                                {
+                                                    if (removedChannel.FunctionInputBuffers.Values.Sum(val => val.Count) < 1 || timer.ElapsedMilliseconds >= _disposeGraceTimeInMilliseconds.TotalMilliseconds)
```

Comment:
add more logging. Use log.debug statements. We need to be able to figure out from Kusto logs, how long we are waiting for invocation buffers to drain